modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java

```diff
@@ -1612,11 +1738,97 @@ private PageMemoryEx getPageMemoryForCacheGroup(int grpId) throws IgniteCheckedE
     }
 
     /**
+     *
+     */
+    private void restoreUpdates(final RecoveryContext recCtx) throws IgniteCheckedException {
+        final Map<T2<Integer, Integer>, T2<Integer, Long>> partStates = new HashMap<>();
```

Comment:
Map of tuples may be too heap-consuming. Check if CachePartitionPartialCountersMap can be used or modified for this use-case.